packages/react-query/src/__tests__/prefetch.test.tsx

```diff
@@ -0,0 +1,352 @@
+
+const generateInfiniteQueryOptions = (
+  data: Array<{ data: string; currentPage: number; totalPages: number }>,
+) => {
+  let currentPage = 0
+
+  return {
+    queryFn: vi
+      .fn<any, Promise<(typeof data)[number]>>()
+      .mockImplementation(async () => {
+        const currentPageData = data[currentPage]
+        if (!currentPageData) {
+          throw new Error('No data defined for page ' + currentPage)
+        }
+
+        await sleep(10)
+        currentPage++
+
+        return currentPageData
+      }),
+    initialPageParam: 1,
+    getNextPageParam: (lastPage: (typeof data)[number]) =>
+      lastPage.currentPage === lastPage.totalPages
+        ? undefined
+        : lastPage.currentPage + 1,
+  }
+}
+
+describe('usePrefetchQuery', () => {
+  const queryCache = new QueryCache()
+  const queryClient = createQueryClient({ queryCache })
+
+  function Suspended<TData = unknown>(props: {
+    queryOpts: UseQueryOptions<TData, Error, TData, Array<string>>
+  }) {
+    const state = useSuspenseQuery(props.queryOpts)
+
+    return (
+      <div>
+        <div>data: {String(state.data)}</div>
+      </div>
+    )
+  }
+
+  it('should prefetch query if query state does not exist', async () => {
+    const queryOpts = {
+      queryKey: queryKey(),
+      queryFn: generateQueryFn('Prefetch is nice!'),
+    }
+
+    function App() {
+      usePrefetchQuery(queryOpts)
+
+      return (
+        <React.Suspense fallback="Loading...">
+          <Suspended queryOpts={queryOpts} />
+        </React.Suspense>
+      )
+    }
+
+    const rendered = renderWithClient(queryClient, <App />)
+
+    await waitFor(() => rendered.getByText('data: Prefetch is nice!'))
+    expect(queryOpts.queryFn).toHaveBeenCalledTimes(1)
+  })
+
+  it('should not prefetch query if query state exists', async () => {
+    const queryOpts = {
+      queryKey: queryKey(),
+      queryFn: generateQueryFn('The usePrefetchQuery hook is smart!'),
+    }
+
+    function App() {
+      usePrefetchQuery(queryOpts)
+
+      return (
+        <React.Suspense fallback="Loading...">
+          <Suspended queryOpts={queryOpts} />
+        </React.Suspense>
+      )
+    }
+
+    await queryClient.fetchQuery(queryOpts)
+    ;(queryOpts.queryFn as Mock).mockClear()
+    const rendered = renderWithClient(queryClient, <App />)
+
+    expect(rendered.queryByText('fetching: true')).not.toBeInTheDocument()
+    await waitFor(() =>
+      rendered.getByText('data: The usePrefetchQuery hook is smart!'),
+    )
+    expect(queryOpts.queryFn).not.toHaveBeenCalled()
+  })
+
+  it('should display an error boundary if query cache is populated with an error', async () => {
+    const queryFn = generateQueryFn('Not an error')
+
+    const queryOpts = {
+      queryKey: queryKey(),
+      queryFn,
+    }
+
+    queryFn.mockImplementationOnce(async () => {
+      await sleep(10)
+
+      throw new Error('Oops! Server error!')
+    })
+
+    function App() {
+      usePrefetchQuery(queryOpts)
```

Comment:
I _think_ to replicate the bug we saw with endless loops, the `usePrefetchQuery` would have to be inside the Suspense boundary so it would rerender->refetch when unsuspended? I haven't tested so not entirely sure though.

Reply:
Hey @Ephem 

What I know about this endless loops bug, is that before it was using the `queryClient.ensureQuery` right? But I will add a test to cover this.

Reply:
@Ephem , I'm not sure, but this endless loops isn't about the usage of `ensureQueryData` instead of `prefetchQuery`? But I will add a test covering the usage of the hook inside the Suspense Boundary.

Reply:
Added new tests to ensure that endless loops don't occur at #7614